### PR TITLE
feat: mostly support Generic

### PR DESCRIPTION
**Pull request by nstarman**
Implements support for Generics.
AI Disclosure: most ideas were mine. I used claude opus with R/G TDD to implement and refine the code. Then Copilot reviews to catch bugs and continue iterating.

I've spent some time optimizing dispatching on generics, but it's still slower than normal types, so bleeding into this PR are some various performance optimizations I added as I saw them. This should also speed up the "normal" dispatch routes on non-parametrized types!

Remaning Qs:
1. Is this behaviour sensible, particularly the fallback dispatch? @PhilipVinc you Julia a lot....
2. is there a beartype mechanism to inspect Generics, preferably a `__method__` that can be fast? @leycec this is for you :).

### CODE REVIEW

**Comment by nstarman** [comment]
CleanShot 2026-05-19 at 20 27 37

It's pretty performant! My naive implementations were ~20x the time. Now it's 1.6
And I sped everything up, so that 1.6 is pretty competitive with the old baseline speed.

**Comment by nstarman** [comment]
@wesselb this got to be a doozy, in a push one corner of the rug another corner pops up kind of way.

File: src/plum/_resolver.py [hunk 1]
```diff
@@ -303,36 +398,138 @@ def register(self, method: Method) -> None:
...
-                )
-
+        # Find the index of the first existing method with an equal signature,
+        # using a generator so we stop at the first match rather than scanning
+        # the full list and building a boolean array.
+        existing_idx = next(
+            (i for i, m in enumerate(self.methods) if m.signature == signature),
+            None,
+        )
+        if existing_idx is not None:
```
[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_resolver.py:420]: WDYT @wesselb ?

File: src/plum/_generic.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+    The class **must** define ``__infer_type_parameter__`` as a
+    :func:`classmethod` that accepts the freshly-constructed instance and
+    returns the type parameter (or a tuple of type parameters for multi-
+    parameter generics).  The decorator raises :exc:`TypeError` at decoration
+    time if the method is absent.
+
+    Subscripted construction (e.g. ``A[str](value)``) still wins: Python sets
+    ``__orig_class__`` *after* ``__init__`` returns, overwriting whatever the
+    wrapper inferred.
```
[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_generic.py]: Fixed.

File: src/plum/_bear.py [hunk 3]
```diff
@@ -1,14 +1,73 @@
...
+    """Return True iff *origin* is a user-defined ``Generic[T]`` subclass.
+
+    Built-in generics (``list``, ``dict``, ``tuple``, etc.) return False because
+    beartype can introspect them directly.  Custom ``class Box(Generic[T])``
+    subclasses return True because beartype cannot verify ``T`` at runtime.
+    """
+    return isinstance(origin, type) and Generic in origin.__mro__
+
+
+def is_bearable_with_orig(value: object, hint: Any, /) -> bool:
```
Comment (nstarman) [inline src/plum/_bear.py:28]: @leycec is there a more beartype way to do this? I hacked this together to also get the type parameter info.

**Comment by nstarman** [comment]
See https://plum-dispatch--272.org.readthedocs.build/en/272/generics.html for the user-facing API.

**Comment by nstarman** [comment]
@wesselb  in a followup I'm interested in unleashing mypyc on a lot of the changes made in this PR. I think there's some major speedups to be had in building better compiled wheels.

File: docs/generics.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+>>> g(Box[str]("hi"))   # __orig_class__ = Box[str] — no str overload → Box fallback
+'bare Box'
+```
+
+## Auto-inferring types with `@plum.generic`
+
+If your class can always infer `T` from the constructor arguments — for
+example, `Box(1)` should behave exactly like `Box[int](1)` — you can opt into
+automatic inference with `@plum.generic`:
```
Comment (nstarman) [inline docs/generics.md]:
```suggestion
If your class can always infer `T` from the constructor arguments — for example, `Box(1)` should behave exactly like `Box[int](1)` — you can opt into automatic inference with `@plum.generic`:
```

File: docs/generics.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+Two scenarios are measured:
+
+- **Faithful** — only a bare `B` overload (no type-parameter overloads).  Plum uses the fast faithful-cache path.
+- **Generic** — `A[Any]`, `A[int]`, and `A[str]` overloads.  Plum uses the two-tier generic cache, keyed on `__orig_class__` when present.
+
+```{include} _generated/generics_timing.md
+```
+
+The faithful path is the fastest because Plum caches by `type(arg)` and checks membership with a single `issubclass` call.  The generic path must additionally read `__orig_class__` (or detect its absence), build the two-tier cache key, and run the TypeHint subtype check on a cache miss.
```
Comment (nstarman) [inline docs/generics.md]:
```suggestion
The faithful path is the fastest because Plum caches by `type(arg)` and checks membership with a single `issubclass` call.  The generic path must additionally read `__orig_class__` (or detect its absence), build a two-tier cache key over the type and type parameters, and run the TypeHint subtype check on a cache miss.
```

File: docs/generics.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+    @classmethod
+    def __infer_type_parameter__(cls, instance):
+        return (type(instance.x), type(instance.y))
+```
+
+`@generic` raises `TypeError` at decoration time if `__infer_type_parameter__` is not defined on the class or any of its ancestors.
+
+```{note}
+`@plum.generic` is a **lightweight** opt-in that simply sets `__orig_class__`. For richer parametric machinery — covariant subclassing, custom type-parameter validation, and multi-parameter ordering — see [Parametric Classes](parametric.md).
```
Comment (nstarman) [inline docs/generics.md]:
```suggestion
`@plum.generic` is a **lightweight** opt-in that sets `__orig_class__`. For richer parametric machinery — covariant subclassing, custom type-parameter validation, and multi-parameter ordering — see [Parametric Classes](parametric.md).
```

File: docs/generics.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+- **Subscripted instances** still pick the most specific overload — `Box[int](1)` prefers `Box[int]` over `Box[Any]`.
+- **Bare instances** match only `Box[Any]` (the other overloads are excluded because there is no parameterization to verify), so dispatch is unambiguous.
+
+```{tip}
+Think of `A[Any]` as the explicit way to say *"this overload is the fallback for any `A` instance whose parameter is unknown at runtime"*.  Without it, bare instances will raise `NotFoundLookupError` when only parameterized overloads are registered.
+```
+
+## Bare class hints (`Box` without brackets)
+
+You can also write a fallback against the **bare** (unsubscripted) class:
```
Comment (nstarman) [inline docs/generics.md]:
```suggestion
You can equivalently write a fallback against the **bare** (unsubscripted) class:
```

**Comment by nstarman** [comment]
Ping @wesselb  :)

**Comment by leycec** [comment]
**...heh.** This is deliciously intense. @beartype does, indeed, provide the public (and admittedly poorly documented) `beartype.door.TypeHint` API. `TypeHint` defines a partial order over the family of *almost* all PEP-compliant type hints – including:

* PEP 484-compliant user-defined generics like `class MuhGeneric[T]():`.
* PEP 585-compliant container type hints like `list[str]`.

Deciding the partial order between a type hierarchy of user-defined generics subscriptable by arbitrary type variables was especially... non-trivial. Merely to remember it, I grit my teeth in mental anguish. Still, it all works reasonably reliably to the best of my knowledge. Let us ignore a spate of recent @beartype issues filed against `beartype.door.is_subhint()` that are surely ignorable. <sup>*surely!*</sup>

Examples or it didn't happen:

```python
from beartype.door import TypeHint

class Pep484GenericST[S, T](): pass
class Pep484GenericSInt[S](Pep484GenericST[S, int]): pass

print(TypeHint(Pep484GenericSInt) <= TypeHint(Pep484GenericST))
print(TypeHint(Pep484GenericSInt) <= TypeHint(Pep484GenericST[int, int]))
print(TypeHint(Pep484GenericSInt[int]) <= TypeHint(Pep484GenericST))
```

...which prints (as thankfully expected):

```python
True
False
True
```

If you'd like even deeper introspection, @beartype can probably do that in theory but probably doesn't do that in practice. It's summertime. I cannot be held responsible for cycling instead of coding. :joy: `->` :sob:

**Comment by wesselb** [comment]
Whoa, @nstarman, this is very very cool. I wasn't aware how far support for generics has come. This PR introduces a lot, but I'm very much in favour! I'll need to take a moment to sit down and go through all details very carefully, but on a skim everything seems to make sense. Is this ready for an in-depth review?

**Comment by nstarman** [comment]
I think so! GH does have some good comments about the topological algorithm, but treating that as a distinct sub-unit, I think the rest is ready.

**Comment by wesselb** [comment]
Fantastic! Let me dig into this. :)

File: src/plum/_resolver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply src/plum/_resolver.py:420]: why are we dropping this? I mean registering should be something done rarely so it's not a huge issue... or is it?

[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_resolver.py:420]: Since `register()` is essentially never on the hot path, I've rolled this part back in c8fde4b.

`register()` now does the full exhaustive scan again and raises if a new signature is equal to more than one existing method, restoring the previous upstream behavior (raised explicitly rather than via `assert`, so it isn't stripped under `-O`):

```python
_equal_indices = [i for i, m in enumerate(self.methods) if m.signature == signature]
if len(_equal_indices) > 1:
    raise AssertionError(
        f"The added method `{method}` is equal to {len(_equal_indices)} "
        f"existing methods. This should never happen."
    )
existing_idx = _equal_indices[0] if _equal_indices else None
```

The `next(...)` first-match early-exit is gone.

The other `register()` change from this PR — updating the derived metadata (`is_faithful`, `is_faithful_for_non_generic`, `generic_origins`, `_arity1_methods`) incrementally instead of rescanning all methods on every call — is kept. That one is a genuine O(N²)→O(N) win over a full registration and is unaffected by this concern: it only consumes the resulting `existing_idx`, and the "REPLACE leaves metadata unchanged" reasoning holds regardless of how the matching index is found (equal signatures ⇒ equal types/arity). Also added a test that asserts the duplicate-signature case raises `AssertionError`.

[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_resolver.py:420]: Addressed in c8fde4b — restored the exhaustive scan plus the fail-fast `AssertionError` when a signature matches more than one existing method (raised explicitly, so it survives `-O`). The incremental metadata updates are kept; only the `next(...)` first-match early-exit was rolled back. See the discussion in the sibling thread.

File: src/plum/_resolver.py [hunk 8]
```diff
@@ -303,36 +398,138 @@ def register(self, method: Method) -> None:
...
                     f"overwrites the earlier definition "
-                    f"`{previous_method}` "
+                    f"`{_replaced_method}` "
                     f"(`{repr_source_path(prev_impl)}`).",
                     category=MethodRedefinitionWarning,
                     stacklevel=0,
                 )
 
-            self.methods[existing.index(True)] = method
+            self.methods[existing_idx] = method
```
[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_resolver.py]: This is now fixed and it does the full O(n) search.

File: src/plum/_resolver.py [hunk 9]
```diff
@@ -231,6 +233,91 @@ def _unwrap_invoked_methods(f: Callable[..., object], /) -> Callable[..., object
...
+    result: list[Method] = []
+    queue = [i for i, d in enumerate(in_degree) if not d]
+    while queue:
+        result.extend(methods[i] for i in queue)
+        next_queue: list[int] = []
+        for i in queue:
+            for j in successors[i]:
+                in_degree[j] -= 1
+                if in_degree[j] == 0:
+                    next_queue.append(j)
```
[the remark this answers is not in the document]
Comment (nstarman) [reply src/plum/_resolver.py:314]: Good catch — the observation is correct. The initial queue was in index order, but each subsequent layer's `next_queue` was built in discovery order (whichever predecessor freed a node last), so incomparable methods reaching in-degree 0 in the same layer could come out reordered — e.g. `[0, 1, 2, 3]` → `[0, 1, 3, 2]`.

To be transparent about severity: this had **no effect on dispatch**. The only consumers are `resolve_for_type` → `_resolve_from`, which recomputes the unique set of most-specific candidates order-independently (and raises `AmbiguousLookupError` / uses precedence for incomparable ties), and the `_generic_cache` prepopulation, which explicitly skips any bucket containing incomparable methods — and the reordering can only occur when incomparable methods share a layer. So it was a docstring-accuracy issue rather than a bug.

Still, since `_sort_most_specific_first` runs only at registration (never on the hot path), making it deterministic is essentially free and worth doing. Applied your suggestion in 82f5908 — added `next_queue.sort()` so every layer is emitted in original-index order, matching the docstring — plus a regression test (`test_sort_most_specific_first_stable_within_layer`) that pins the `[0, 1, 2, 3]` case.

**Comment by nstarman** [comment]
@wesselb, sorry for the delay, I've been traveling. I've now addressed the PR review comments — fixing ordering in the resolver and rolling back the dangerous O(1) registration for the safer O(n) original implementation.

**Comment by nstarman** [comment]
@wesselb, sorry for the delay, I've been traveling. I've now addressed the PR review comments — fixing ordering in the resolver and rolling back the dangerous O(1) registration for the safer O(n) original implementation.

**Comment by wesselb** [comment]
Thanks for the updates, @nstarman! I need to familiarise myself again with what's going on here.

On a skim, it seems that there are a variety of optimisations to improve performance. The PR is pretty large and complicated. Do you think all optimisations are necessary? I'd be happy to sacrifice a little bit of performance if that would make the implementation substantially simpler.

**Comment by nstarman** [comment]
Most of the diff size is for implementing and testing Generic support (>80% by my line count, though that includes misc cleanups like `Function` -> `plum.Function` in tests).
A vanilla implementation of Generic support is quite slow, but the 2-tiered cache is much more performant, which is why the Generic-related part of the diff is quite large.
Even so, without further optimizations and fast-paths, a dispatch on a Generic is around 2x slower than on other types. 
To simplify this PR we could try implementing the other performance improvements first, rebasing this PR as necessary ?

**Comment by leycec** [comment]
@leycec likes what he sees here despite understanding nothing that he sees here. From the little that I understand, a 200% speedup for multiple dispatch on generics seems extraordinarily worthwhile. That's not a negligible microoptimization – unlike most of what I do in the @beartype codebase. :joy: 

You've done a sterling job here, @nstarman. This must be why one of us gets prizes and works at MIT. :clap: :clap: 

If anyone's curious, here's the tentative @beartype `0.23.0` release timetable:

* Finish resolving beartype/beartype#556 *and* beartype/beartype#670 next week, both of which I consider hard blockers for release. Both are mostly done. Mostly. That adverb sure is a weasel word, isn't it?
* **Release @beartype `0.23.0rc0` Release Candidate 0 next Friday** (June 24th, apparently). Gods. Where is this summer even going!? Emoji cat ain't happy. :crying_cat_face: 
* **Release @beartype `0.23.0` stable next...** uhh, no idea. August 7th, maybe? Let's pretend that means something.

Thus spake Bearathustra. What a QA guy, that Bearathustra.

**Comment by nstarman** [comment]
Oh no. perpetual merge conflicts. I need to stop doing other PRs :p

**Comment by nstarman** [comment]
@wesselb this PR was informative, but #290 identifies an approach that is many times faster, simpler, and more modular/extensible. I've marked this PR as a draft.
